src/runtime/library_module.cc

```diff
@@ -0,0 +1,165 @@
+ */
+class ModuleInternal {
+ public:
+  // Get mutable reference of imports.
+  static std::vector<Module>* GetImportsAddr(ModuleNode* node) {
+    return &(node->imports_);
+  }
+};
+
+PackedFunc WrapPackedFunc(BackendPackedCFunc faddr,
+                          const ObjectPtr<Object>& sptr_to_self) {
+  return PackedFunc([faddr, sptr_to_self](TVMArgs args, TVMRetValue* rv) {
+      int ret = (*faddr)(
+          const_cast<TVMValue*>(args.values),
+          const_cast<int*>(args.type_codes),
+          args.num_args);
+      CHECK_EQ(ret, 0) << TVMGetLastError();
+    });
+}
+
+void InitContextFunctions(std::function<void*(const char*)> fgetsymbol) {
+  #define TVM_INIT_CONTEXT_FUNC(FuncName)                          \
+    if (auto *fp = reinterpret_cast<decltype(&FuncName)*>          \
+        (fgetsymbol("__" #FuncName))) {                            \
+      *fp = FuncName;                                              \
+    }
+  // Initialize the functions
+  TVM_INIT_CONTEXT_FUNC(TVMFuncCall);
+  TVM_INIT_CONTEXT_FUNC(TVMAPISetLastError);
+  TVM_INIT_CONTEXT_FUNC(TVMBackendGetFuncFromEnv);
+  TVM_INIT_CONTEXT_FUNC(TVMBackendAllocWorkspace);
+  TVM_INIT_CONTEXT_FUNC(TVMBackendFreeWorkspace);
+  TVM_INIT_CONTEXT_FUNC(TVMBackendParallelLaunch);
+  TVM_INIT_CONTEXT_FUNC(TVMBackendParallelBarrier);
+
+  #undef TVM_INIT_CONTEXT_FUNC
+}
+
+/*!
+ * \brief Load and append module blob to module list
+ * \param mblob The module blob.
+ * \param module_list The module list to append to
+ */
+void ImportModuleBlob(const char* mblob, std::vector<Module>* mlist) {
+#ifndef _LIBCPP_SGX_CONFIG
+  CHECK(mblob != nullptr);
+  uint64_t nbytes = 0;
+  for (size_t i = 0; i < sizeof(nbytes); ++i) {
+    uint64_t c = mblob[i];
+    nbytes |=  (c & 0xffUL) << (i * 8);
+  }
+  dmlc::MemoryFixedSizeStream fs(
+      const_cast<char*>(mblob + sizeof(nbytes)), static_cast<size_t>(nbytes));
+  dmlc::Stream* stream = &fs;
+  uint64_t size;
+  CHECK(stream->Read(&size));
+  for (uint64_t i = 0; i < size; ++i) {
+    std::string tkey;
+    CHECK(stream->Read(&tkey));
+    std::string fkey = "module.loadbinary_" + tkey;
+    const PackedFunc* f = Registry::Get(fkey);
+    CHECK(f != nullptr)
+        << "Loader of " << tkey << "("
+        << fkey << ") is not presented.";
+    Module m = (*f)(static_cast<void*>(stream));
+    mlist->push_back(m);
+  }
+#else
+  LOG(FATAL) << "SGX does not support ImportModuleBlob";
+#endif
+}
+
+Module CreateModuleFromLibrary(ObjectPtr<Library> lib) {
+  InitContextFunctions([lib](const char* fname) {
+      return lib->GetSymbol(fname);
+    });
+  auto n = make_object<LibraryModuleNode>(lib);
+  // Load the imported modules
+  const char* dev_mblob =
+      reinterpret_cast<const char*>(
+          lib->GetSymbol(runtime::symbol::tvm_dev_mblob));
+  if (dev_mblob != nullptr) {
+    ImportModuleBlob(
+        dev_mblob, ModuleInternal::GetImportsAddr(n.operator->()));
+  }
+
+  Module root_mod = Module(n);
```

Comment:
Suggest changing `root_mod` into `dso_module`? Here, n is DSOModuleLib. Because I think of my WIP runtime implementation, our dso module's root module maybe is `graph_runtime` / `vm` and so on. If we name it `root_mod`, which will confuse the concept in the future we change it.

Reply:
Indeed we want to return the root module, and set the context addr to the root module(so any child class can get functions from other modules in the hierachy)

Reply:
but doesn’t it be ‘root_module->import(rood_mod)’? root_module is GraphRuntime / VM, root_mod is the dso module here which provides other symbol so that others could access. Am I misunderstand the concept here? Please correct me. Thanks.

Reply:
We will work fine as GetFuncFromEnv also get the symbols recursively(which means even if root_module is Graphruntime, it should be able to fine these symbols)